### PR TITLE
update related manifest files to use hostpath type

### LINKED ISSUE #298: Change kube-proxy xtables container to use a file type mount
## Versions

**kubeadm version** (use `kubeadm version`):
master

**Environment**:
- **Kubernetes version** (use `kubectl version`): 1.7

At the moment we rely on kubernetes/kubernetes#47212 to create the xtables file upon kubelet startup. This means that we don't need anymore prep when mounting the xtables path into the kube-proxy. Since kubernetes/kubernetes#47212 has a TODO about fixing init-containers, we cannot rely on it to be there in future releases. 

When kubernetes/kubernetes#46597 merges, we should change the xtables lock to a file based mount. This should create the xtables lock as a file if it does not already exist

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:
Per [discussion in #46597](https://github.com/kubernetes/kubernetes/pull/46597#pullrequestreview-53568947)

Dependes on #46597

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

Fixes: https://github.com/kubernetes/kubeadm/issues/298

**Special notes for your reviewer**:
/cc @euank @thockin @tallclair @Random-Liu 

**Release note**:

```release-note
None
```

### CODE REVIEW

**Comment by thockin** [comment]
Needs rebase.  Ping me when done.

**Comment by dixudx** [comment]
@luxas Seems the root cause is `HostPathType` that cause [pull-kubernetes-unit](https://k8s-gubernator.appspot.com/builds/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-unit) failed. I'm working on the patch and testing on local env. Will submit after finishing testing.

**Comment by luxas** [comment]
@dixudx Please include the change for https://github.com/kubernetes/kubeadm/issues/298 as well

**Comment by msau42** [comment]
Shouldn't the correct fix be to make the new field optional?

**Comment by thockin** [comment]
The field is optional.  This was not intended to fix any error, just to use a new feature.

File: cmd/kubeadm/app/phases/controlplane/volumes.go [hunk 1]
```diff
@@ -60,23 +62,23 @@ func getHostPathVolumesForTheControlPlane(cfg *kubeadmapi.MasterConfiguration) c
...
+	mounts.NewHostPathMount(kubeadmconstants.KubeControllerManager, kubeadmconstants.KubeCertificatesVolumeName, cfg.CertificatesDir, cfg.CertificatesDir, true, &hostPathDirectoryOrCreate)
 	// Read-only mount for the ca certs (/etc/ssl/certs) directory
-	mounts.NewHostPathMount(kubeadmconstants.KubeControllerManager, caCertsVolumeName, caCertsVolumePath, caCertsVolumePath, true)
+	mounts.NewHostPathMount(kubeadmconstants.KubeControllerManager, caCertsVolumeName, caCertsVolumePath, caCertsVolumePath, true, &hostPathDirectoryOrCreate)
 	// Read-only mount for the controller manager kubeconfig file
 	controllerManagerKubeConfigFile := filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.ControllerManagerKubeConfigFileName)
-	mounts.NewHostPathMount(kubeadmconstants.KubeControllerManager, kubeadmconstants.KubeConfigVolumeName, controllerManagerKubeConfigFile, controllerManagerKubeConfigFile, true)
+	mounts.NewHostPathMount(kubeadmconstants.KubeControllerManager, kubeadmconstants.KubeConfigVolumeName, controllerManagerKubeConfigFile, controllerManagerKubeConfigFile, true, &hostPathFileOrCreate)
```
Comment (luxas) [inline cmd/kubeadm/app/phases/controlplane/volumes.go]: minor nit: This should probably be `hostPathFile`. If the file doesn't exist, the best behavior is probably to make the kubelet fail; as no credentials to the API server is a fatal error

File: cmd/kubeadm/app/phases/controlplane/volumes.go [hunk 2]
```diff
@@ -42,14 +42,16 @@ var caCertsPkiVolumePath = "/etc/pki"
 
 // getHostPathVolumesForTheControlPlane gets the required hostPath volumes and mounts for the control plane
 func getHostPathVolumesForTheControlPlane(cfg *kubeadmapi.MasterConfiguration) controlPlaneHostPathMounts {
+	hostPathDirectoryOrCreate := v1.HostPathDirectoryOrCreate
```
Comment (luxas) [inline cmd/kubeadm/app/phases/controlplane/volumes.go]: nit: why declare temporary variables?

**Review by luxas — APPROVED** [review]
kubeadm changes LGTM
/lgtm

Also verified manually that it works for v1.7 clusters as well

Thanks!

File: cmd/kubeadm/app/phases/controlplane/volumes.go (hunk 2 above)
Comment (dixudx) [reply cmd/kubeadm/app/phases/controlplane/volumes.go]: The `HostPathType` is supposed to be a pointer. We have to declare such temporary variables.

File: cmd/kubeadm/app/phases/controlplane/volumes.go (hunk 1 above)
Comment (dixudx) [reply cmd/kubeadm/app/phases/controlplane/volumes.go]: Just wanna to make it more tolerant. @luxas Change it to `File`?

**Pull request merged**